### PR TITLE
Use leader election library from csi-lib-utils 

### PR DESCRIPTION
**Pull request by andrewsykim**
Also adds support for lease based leader election which is highly recommended over endpoints base. Endpoints based will be supported for a few releases for backwards compatibility. 

/assign @msau42 @jsafrane  @xing-yang

### CODE REVIEW

**Comment by andrewsykim** [comment]
/hold 

Waiting for https://github.com/kubernetes-csi/csi-lib-utils/pull/24

**Comment by pohly** [comment]
How does this affect RBAC rules?

File: cmd/csi-provisioner/csi-provisioner.go [hunk 1]
```diff
@@ -52,19 +54,25 @@ var (
...
-	enableLeaderElection = flag.Bool("enable-leader-election", false, "Enables leader election. If leader election is enabled, additional RBAC rules are required. Please refer to the Kubernetes CSI documentation for instructions on setting up these RBAC rules.")
 	retryIntervalStart   = flag.Duration("retry-interval-start", time.Second, "Initial retry interval of failed provisioning or deletion. It doubles with each failure, up to retry-interval-max.")
 	retryIntervalMax     = flag.Duration("retry-interval-max", 5*time.Minute, "Maximum retry interval of failed provisioning or deletion.")
 	workerThreads        = flag.Uint("worker-threads", 100, "Number of provisioner worker threads, in other words nr. of simultaneous CSI calls.")
 	operationTimeout     = flag.Duration("timeout", 10*time.Second, "Timeout for waiting for creation or deletion of a volume")
 	_                    = deprecatedflags.Add("provisioner")
 
+	enableLeaderElection = flag.Bool("enable-leader-election", false, "Enables leader election. If leader election is enabled, additional RBAC rules are required. Please refer to the Kubernetes CSI documentation for instructions on setting up these RBAC rules.")
```
Comment (msau42) [inline cmd/csi-provisioner/csi-provisioner.go]: Add deprecation notice here

File: cmd/csi-provisioner/csi-provisioner.go [hunk 2]
```diff
@@ -178,8 +186,31 @@ func init() {
...
+	if !*enableLeaderElection {
+		run(context.TODO())
+	} else {
+		// this lock name pattern is also copied from sigs.k8s.io/sig-storage-lib-external-provisioner/controller
+		// to preserve backwards compatibility
+		lockName := strings.Replace(provisionerName, "/", "-", -1)
+
+		var le leaderElection
+		if *leaderElectionType == "endpoints" {
+			le = leaderelection.NewLeaderElectionWithEndpoints(clientset, lockName, run)
```
Comment (msau42) [inline cmd/csi-provisioner/csi-provisioner.go]: And deprecation log message here

File: deploy/kubernetes/rbac.yaml [hunk 3]
```diff
@@ -79,6 +79,9 @@ rules:
 - apiGroups: [""]
   resources: ["endpoints"]
   verbs: ["get", "watch", "list", "delete", "update", "create"]
+- apiGroups: ["coordination.k8s.io"]
```
Comment (msau42) [inline deploy/kubernetes/rbac.yaml]: And a comment that only one of endpoints or lease rules is needed

File: cmd/csi-provisioner/csi-provisioner.go (hunk 1 above)
Comment (andrewsykim) [reply cmd/csi-provisioner/csi-provisioner.go]: Did you mean for `leaderElectionType`?

**Pull request merged**